pennylane/templates/subroutines/prepselprep.py

```diff
@@ -0,0 +1,209 @@
+        new_op = qml.simplify(qml.ops.LinearCombination([sign], [op]))
+        new_ops.append(new_op)
+
+        sign = qml.math.sign(imag)
+        new_coeffs.append(sign * imag)
+        new_op = qml.simplify(qml.ops.LinearCombination([1j * sign], [op]))
+        new_ops.append(new_op)
+
+    return new_coeffs, new_ops
+
+
+def new_terms_is_real(lcu):
+    """Computes new terms when the coefficients are real.
+    This preserves the number of terms."""
+
+    new_coeffs = []
+    new_ops = []
+    for coeff, op in zip(*lcu.terms()):
+        sign = qml.math.sign(coeff)
+        new_coeffs.append(sign * coeff)
+        new_op = qml.simplify(qml.ops.LinearCombination([sign], [op]))
+        new_ops.append(new_op)
+
+    return new_coeffs, new_ops
+
+
+class PrepSelPrep(Operation):
+    """This class implements a block-encoding of a linear combination of unitaries
+    using the Prepare, Select, Prepare method"""
+
+    def __init__(self, lcu, control=None, id=None):
+        coeffs, ops = lcu.terms()
+        control = qml.wires.Wires(control)
+        self.hyperparameters["lcu"] = lcu
+        self.hyperparameters["coeffs"] = coeffs
+        self.hyperparameters["ops"] = ops
+        self.hyperparameters["control"] = control
+
+        if any(
+            control_wire in qml.wires.Wires.all_wires([op.wires for op in ops])
+            for control_wire in control
+        ):
+            raise ValueError("Control wires should be different from operation wires.")
+
+        target_wires = qml.wires.Wires.all_wires([op.wires for op in ops])
+        self.hyperparameters["target_wires"] = target_wires
+
+        all_wires = target_wires + control
+        super().__init__(*self.data, wires=all_wires, id=id)
+
+    def _flatten(self):
+        return tuple(self.lcu), (self.control)
+
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "PrepSelPrep":
+        coeffs = [term.terms()[0][0] for term in data]
+        ops = [term.terms()[1][0] for term in data]
+        lcu = qml.ops.LinearCombination(coeffs, ops)
+        return cls(lcu, metadata)
+
+    def __repr__(self):
+        return f"PrepSelPrep(coeffs={tuple(self.coeffs)}, ops={tuple(self.ops)}, control={self.control})"
+
+    def map_wires(self, wire_map: dict) -> "PrepSelPrep":
+        new_ops = [o.map_wires(wire_map) for o in self.hyperparameters["ops"]]
+        new_control = [wire_map.get(wire, wire) for wire in self.hyperparameters["control"]]
+        new_lcu = qml.ops.LinearCombination(self.hyperparameters["coeffs"], new_ops)
+        return PrepSelPrep(new_lcu, new_control)
+
+    def decomposition(self):
+        return self.compute_decomposition(self.lcu, self.control)
+
+    @staticmethod
+    def normalization_factor(lcu):
+        """Return the normalization factor lambda such that
+        A/lambda is in the upper left of the block encoding"""
+
+        new_coeffs, _ = get_new_terms(lcu)
+        return qml.math.sum(new_coeffs)
```

Comment:
Why this is needed? Please remove if not necessary.

Reply:
This was used in the block encoding test. I can get rid of it and just change the test.